Ensure config map is not created too early

kajinamit opened the pull request:
This introduces a few more assertions to ensure config map is created unless the required input and Keystone API is ready.

Depends-on: https://github.com/openstack-k8s-operators/lib-common/pull/294

SeanMooney commented on controllers/placementapi_controller.go:
```diff
@@ -31,7 +31,6 @@ import (
 	"github.com/go-logr/logr"
 	routev1 "github.com/openshift/api/route/v1"
 	keystonev1 "github.com/openstack-k8s-operators/keystone-operator/api/v1beta1"
```
nit: you could have kept this as just keystone

stuggi replied:
its common when using APIs to name it using the major version.

SeanMooney replied:
ack its fine either way i just called it out as you could have avoided renameing it below but if we are only gong to rev that when we rev a major version and not everytime we change a minor version I'm totally fine with that

SeanMooney commented on controllers/suite_test.go:
```diff
@@ -67,30 +71,88 @@ func TestAPIs(t *testing.T) {
 }
 
-func GetDependencyVersion(moduleName string) (string, error) {
+func GetDependencyVersion(moduleName string) (string, string, error) {
```
its too bad these return type are not named
in isolation its impossible to know what the first and second return argument
are without treading the body fo the function 
 name, version, nil

SeanMooney replied:
looking at https://www.geeksforgeeks.org/named-return-parameters-in-golang/
we are allowed to name them so please do.

i.e.

func add(x int, y int) (int) {
	return x + y
}


func add(x int, y int) (ret int) {
	return x + y
}

func add(x int, y int) (ret int) {
    ret = x+y
	return
}

are all legal and do the same thing but the second form is the most readble

so we should use 
func GetDependencyVersion(moduleName string) (name string, version string, err error) {

stuggi replied:
ack will do

stuggi replied:
done

SeanMooney replied:
thanks

SeanMooney reviewed (approved):
this looks good to me but lets see what dan and gibi think

gibizer commented:
I have concerns:
- 1) I'm not sure that a version number from keystone/api will always exists for keystone. Imagine if we only fix something in the API definition (i.e. api documentation) and release that as v1.0.1 from keystone/api but as it does not required any implementation change the latest keystone module version is kept as v1.0.0. So this approach will only work if we always release keystone/api and keystone together.
- 2) This is now a lot of hoops to jump to get something that we could be easily get by storing the API Structs and the generated CRD at the same module.
- 3) Following from 2) If we store the structs that defines the CRD in keystone/api then I think we should also store the CRD that is  directly generated from those structs in the same module. As the API structs also the CRD is part of the API definition.

SeanMooney reviewed (dismissed):
i dont know how to change approvals but this only had the first 3 commits when i reviewed it last
---- latere
no that's not right instead fo rebasing and fixing my comment in the commit martin added the 4th commit and i just didn't see that in the UI because github sucks for review.

SeanMooney commented:
https://stackoverflow.com/a/72813625

i dont like that if you change the pr after you review the approval flag is not cleard

SeanMooney commented:
> I have concerns:
> 
>     * 1. I'm not sure that a version number from keystone/api will always exists for keystone. Imagine if we only fix something in the API definition (i.e. api documentation) and release that as v1.0.1 from keystone/api but as it does not required any implementation change the latest keystone module version is kept as v1.0.0. So this approach will only work if we always release keystone/api and keystone together.
> 
>     * 2. This is now a lot of hoops to jump to get something that we could be easily get by storing the API Structs and the generated CRD at the same module.
> 
>     * 3. Following from 2) If we store the structs that defines the CRD in keystone/api then I think we should also store the CRD that is  directly generated from those structs in the same module. As the API structs also the CRD is part of the API definition.

so i obviously misunderstood the first commit
i thought we were going to go with the symlinking approach and this was going to use the new version of the keystone operator where that had already been done?

and that is why we were getting the CRD form the API module

stuggi commented:
> > I have concerns:
> > ```
> > * 1. I'm not sure that a version number from keystone/api will always exists for keystone. Imagine if we only fix something in the API definition (i.e. api documentation) and release that as v1.0.1 from keystone/api but as it does not required any implementation change the latest keystone module version is kept as v1.0.0. So this approach will only work if we always release keystone/api and keystone together.

This is a valid concern and yes requires both versions stay in sync

> > 
> > * 2. This is now a lot of hoops to jump to get something that we could be easily get by storing the API Structs and the generated CRD at the same module.
> > 
> > * 3. Following from 2) If we store the structs that defines the CRD in keystone/api then I think we should also store the CRD that is  directly generated from those structs in the same module. As the API structs also the CRD is part of the API definition.
> > ```
> 
> so i obviously misunderstood the first commit i thought we were going to go with the symlinking approach and this was going to use the new version of the keystone operator where that had already been done?

No, this PR was to demonstrate an optional way to instead of using the linked CRDs to load the base module with the information of the api mod from go.mod. Sorry if that was not clear.

stuggi commented:
storing in the CRDs in /api and add a symlink from the original place to be able to use the default create mechanism is the easiest and cleanest way.

stuggi commented:
I'll get a new one filed now that we have https://github.com/openstack-k8s-operators/keystone-operator/pull/102 and https://github.com/openstack-k8s-operators/mariadb-operator/pull/35

kajinamit committed:
Ensure config map is not created too early

This introduces a few more assertions to ensure config map is created
unless the required input and Keystone API is ready.

Depends-on: https://github.com/openstack-k8s-operators/lib-common/pull/294

gibizer reviewed (approved):
Nice test coverage improvement. Thanks Takashi! As this is test only I'm OK to merge this before the dev preview cut.
